text/000-wasm-bindgen-trait-method-inheritance.md

```diff
@@ -0,0 +1,401 @@
+following syntax:
+
+```rust
+#[wasm_bindgen]
+pub trait IEventTarget {
+    #[wasm_bindgen(js_name = addEventListener)]
+    fn add_event_listener(&self, name: &str, listener: &Closure<FnMut(Event)>, use_capture: bool);
+
+    #[wasm_bindgen(js_name = removeEventListener)]
+    fn remove_event_listener(&self, name: &str, listener: &Closure<FnMut(Event)>, use_capture: bool);
+
+    #[wasm_bindgen(js_name = dispatchEvent)]
+    fn dispatch_event(&self, event: Event);
+}
+```
+
+```rust
+#[wasm_bindgen]
+pub trait INode: IEventTarget {
+    #[wasm_bindgen(getter = nodeName)]
+    fn node_name(&self) -> JsString;
+
+    #[wasm_bindgen(getter = textContent)]
+    fn text_content(&self) -> JsString;
+
+    #[wasm_bindgen(js_name = appendChild)]
+    fn append_child<A: INode>(&self, child: A) -> A;
+
+    #[wasm_bindgen(js_name = removeChild)]
+    fn remove_child<A: INode>(&self, child: A) -> A;
+
+    // And a bunch more methods...
+}
+```
+
+When the `wasm-bindgen` attribute is used on a trait, it will do two things:
+
+1. It adds an `AsRef<JsValue>` constraint, so `trait Foo` becomes `trait Foo: AsRef<JsValue>`
+
+2. It generates externs for each method, and rewrites the methods so that they have default implementations which call the externs.
+
+The end result is that the above two traits become translated into this:
+
+```rust
+#[wasm_bindgen]
+extern {
+    #[wasm_bindgen(method, structural, js_name = addEventListener)]
+    fn add_event_listener(this: &JsValue, name: &str, listener: &Closure<FnMut(Event)>, use_capture: bool);
+
+    #[wasm_bindgen(method, structural, js_name = removeEventListener)]
+    fn remove_event_listener(this: &JsValue, name: &str, listener: &Closure<FnMut(Event)>, use_capture: bool);
+
+    #[wasm_bindgen(method, structural, js_name = dispatchEvent)]
+    fn dispatch_event(this: &JsValue, event: Event);
+}
+
+pub trait IEventTarget: AsRef<JsValue> {
+    #[inline]
+    fn add_event_listener(&self, name: &str, listener, &Closure<FnMut(Event)>, use_capture: bool) {
+        add_event_listener(self.as_ref(), name, listener, use_capture)
+    }
+
+    #[inline]
+    fn remove_event_listener(&self, name: &str, listener: &Closure<FnMut(Event)>, use_capture: bool) {
+        remove_event_listener(self.as_ref(), name, listener, use_capture)
+    }
+
+    #[inline]
+    fn dispatch_event(&self, event: Event) {
+        dispatch_event(self.as_ref(), event)
+    }
+}
+```
+
+```rust
+#[wasm_bindgen]
+extern {
+    #[wasm_bindgen(method, structural, getter = nodeName)]
+    fn node_name(this: &JsValue) -> JsString;
+
+    #[wasm_bindgen(method, structural, getter = textContent)]
+    fn text_content(this: &JsValue) -> JsString;
+
+    #[wasm_bindgen(method, structural, js_name = appendChild)]
+    fn append_child<A: INode>(this: &JsValue, child: A) -> A;
+
+    #[wasm_bindgen(method, structural, js_name = removeChild)]
+    fn remove_child<A: INode>(this: &JsValue, child: A) -> A;
```

Comment:
This also relates a bit to method overloading (which occurs at least for constructors), in that the generic argument here is really serving as a way to handle multiple signatures. I would say that any generic bindings should be discussed as a future extension to the bindings. So for now I would use the following signature which should be forward-compatible with @fitzgen suggestion:
```rust
pub trait INode: AsRef<Node> + IEventTarget {
    fn append_child(&self, child: &Node) {
        // ...
    }
    // ...
}
```

Reply:
> I don't think extern functions can be generic.

To be clear, the generic is on an `extern` which is annotated with `wasm_bindgen`, so wasm-bindgen could replace the generic `extern` with a non-generic `extern`.

For example, this:

```rust
#[wasm_bindgen]
extern {
    fn foo<A: Foo>(A) -> A;
}
```

...can be compiled into something like this:

```rust
#[wasm_bindgen]
extern {
    fn __generated_foo(JsValue) -> JsValue;
}

#[inline]
fn foo<A: Foo>(value: A) -> A {
    __generated_foo(value.into()).unchecked_into()
}
```

(Where `Foo` extends from `JsCast`)

Reply:
@ohanar The purpose of the generic with `append_child` is to retain type information, because the `append_child` method always returns the argument you pass in, so if you pass in an `HTMLElement` you get back an `HTMLElement`.

Having it instead always return a `Node` causes a loss of type information, which can only be re-obtained by using `dyn_into`, with some runtime cost and loss of ergonomics.

Reply:
So two things here.
- First is that there is no way to know that what more specific return type you have on a method if you have a more specific input type by examining the WebIDL. Consider the following snippet:
```
interface Node : EventTarget {
    Node appendChild(Node node);
    ...
};
```
There is no way to know that if you actually pass an `HTMLElement` into the `appendChild` method you get back an `HTMLElement`. This could be addressed by adding our own custom attribute to the operation (e.g. `WbgReturnType=typeof(node)`), but that would require us to comb through many files and add such annotation (in addition to the work involved in codegen).
- Secondly, the return value for this particular method is actually pretty useless. Interface objects are passed by (shared) reference, not by value. In the case that you pass in an `HTMLElement`, you would still have complete access to the original object after the method call, meaning that you don't need the return value to have access to that particular `HTMLElement`.

Reply:
> First is that there is no way to know that what more specific return type you have on a method if you have a more specific input type by examining the WebIDL. Consider the following snippet:

That's a really good point! It's true that in practice the `appendChild` method always returns its argument, but that isn't actually specified in the WebIDL.

That's a good argument for avoiding the generics for now, since in my opinion the Rust bindings should be close to the WebIDL bindings.

> Secondly, the return value for this particular method is actually pretty useless.

Yes, but according to the WebIDL it *does* return something, so unless you add in a custom attribute it will generate a return value automatically.

Reply:
> Yes, but according to the WebIDL it does return something, so unless you add in a custom attribute it will generate a return value automatically.

Completely agree, it should still return a `Node`, and we shouldn't change that (regardless of how useful it is).